### ISSUE TITLE
Docs and talks: explain why streams/observables

### ISSUE THREAD

**Issue by 3n-mb** [cyclejs/cyclejs#693]
### Problem
New people don't see value in using streams/observables as a primary structuring of app's logic.
New people are also told that this is complicated, and they have to learn observables first. And, as far as I checked, all tutorials start with "this is complicated" -- not a good priming when you want to teach subject, as simple as observables.

What if instead of scaring pants out of newbies, we present core concepts in a way they can relate right away. And use it both in docs and talks (especially talks).

#### Core awesomeness
Do you know events and callbacks? These are underlining everything in web app. They make several logical processes to flow asynchronously to each other, allowing for a reach application experience.

Yet, events and callbacks are a mess to deal with directly. Wouldn't it be nice, if this mess is abstracted in some way? You are lucky. Some comp-sci folks created an easy-for-use abstraction and it has been solidly implemented. Here it is.

Let's take a click event. It is one type of events, and there can be many clicks, producing many events of the same type. Let's think of events as first order objects. One abstract layer up is a stream of events, a second order object.

Like plumber composes plumbing of pipe sections, like electric engineer composes big electric circuits from small sub-section circuits, we too can change, merge, branch and shape flows of events, using chain-able operators on streams (observables).

When you use streams, you explicitly shape your process/data flow. Any spaghetti-like mess becomes immediately apparent.
Explicit view of a data/process flow will allow you to imagine additional processing that cannot be seen when dealing only directly with events, first order objects. Saying "Not seeing the forest behind the trees" is literally applicable here. First order objects, events, are trees. And tons of usual code needed to work with events directly robs an ability to see an overall process (the forest).

When you clearly articulate application processes, the breadth of opportunities opens up. Most of tasks for shaping asynchronous processes can be handled by stream operators. You can debounce, you may sync two streams (zip operator), you can use only some events. And, as a cherry on a cake, setup and cleanup are trivial.

Try it (streams/observable), you'll love it. Money-back guarantee :smile: 

#### Cycle
Cycle is a UI framework that takes streams to its heart. All processes in cycle are shaped using second order objects, streams. Cycle app `main` function produces a chain of streams, a piping for app's events, a piping for app's processes.

Due to use of second order objects, streams, and due to use of stream operators that cleanly implement hairy process/flow shaping/management for you, your code will mostly have functions that do only your application's business logic. Your code will tend to shape into what some call functional style programming. And since use of events is a reactive programming, when you use cycle, you can proudly tell your audience/client/boss that you are doing a glorious *Functional Reactive Programming*.

===== End of sample =====

 - What do you think?
 - What would you add, remove, simplify?

Do use this. Essentially, this is an outline of my onboarding people to observables. And it worked for me :nerd_face:

**Comment by geovanisouza92** [comment]
The most important issue that I saw when trying to teach people about Cycle.js and Streams in general, is the mindset shift needed for people go from A (imperactive) to B (FRP).

People are used to just hang code one on top of another, just like a Christmas tree: You get the button with `document.queryElement...`, then you attach a listener (`.on...` or `.addEventListener`) and put some logic that validate something, starts a HTTP request, then replace update something on the screen. React and all componentization frameworks, just compartimentalize that pattern into smaller pieces, but that pattern is still in there. Some frameworks (and even languages like Elm) tries to warn "hey, try to avoid this... why not to try something like this...", but the user still have the power to create monkey-patches all over the place.

Cycle.js on the other hand, is opinionated and forces a specific kind of coding that, in the longer term, could proves worthy, but is hard to see at first. Like many talks and the docs, we all say "a Cycle.js app is just a component, composed of components. The pattern is the same everywhere", but people are used to expect specific kinds of patterns/structure on determined places.

---

About Streams I think that a nice way to explain it is to pick a commonly seen code snippet (maybe the example I gave above), slice it, and drag parts here and there, allowing people see that the event/listener pattern is still there, but instead of thinking A (event) to B (listener), they need to think B (who is interested?) to A (in what?).

---

I didn't agree (respectfully) with you on saying that Cycle.js is an _UI library_. Besides that was the first objetive when @staltz started writing it, it became much more an *architectural pattern*. The proof is that are [many](https://gist.github.com/geovanisouza92/0f73a5787117e82a26e37c4b5e463464#file-server-js) [examples](https://github.com/geovanisouza92/cycle-lambda) that even don't use DOM driver at all.

Cycle.js (at least `@cycle/run*`) could be used everywhere to structure asynchronous streams graphs, IMO.

**Comment by 3n-mb** [comment]
@geovanisouza92 

> I didn't agree (respectfully) with you on saying that Cycle.js is an UI library. 

I do Ctrl+F with word "library", and it is not in sample. I guess, you are recalling my own views from #579 . Notice how I didn't put them into what I suggest for cycle community as talking points.
Let's quote my text again:

> Cycle is a UI framework that takes streams to its heart. All processes in cycle are shaped using second order objects, streams. Cycle app main function produces a chain of streams, a piping for app's events, a piping for app's processes.

So, :smile: , I respectfully ask for retraction of your comment.

**Comment by geovanisouza92** [comment]
Oh, it was a typo. I should write "UI framework", sorry.

Anyway, I hold my opinion about seeing Cycle.js just for UI, it's far way beyond that. You could use WS (driver) + onionify on server-side without any UI involved. Ok, it shines at UI, but not restricted to it.

**Comment by 3n-mb** [comment]
@geovanisouza92 

> I hold my opinion about seeing Cycle.js just for UI, it's far way beyond that.

The *far way beyond* part is structuring process with observables/streams.
If you associate this with cycle only, and project such view onto newbies, then they will not be writing observables-based, or cycle style logic, when they are not using cycle.

Very recently I wrote a, what you may call, cycle-style app that uses AngularJS with intent sources and view commands sinks. That's right. It is angular 1.x, not even angular n.
And from this experience I am saying that structuring process with observables/streams can be done even if cycle framework is not used.

This mix of concepts reminds me of the following. Use of meditation doesn't require you to be buddhist or to believe any gods. In fact, without gods, meditation tool is way more efficient.
So, let's not push people into thinking that structuring process with observables/streams requires use of cycle.

Now its your tern to ask, why didn't I use cycle? No templates, and, as a natural outcome, no ready material components library. That's why my team couldn't use cycle. And I bet that 90% will be like us.
Yet, we are capable of structuring process with observables/streams, a cycle style.
I'd argue that it is easier to write cycle style in cycle. So, ask yourself, why do we bother to write this style outside of helping cycle environment. My answer is: structuring process with observables/streams.

**Comment by 3n-mb** [comment]
@geovanisouza92 if there is other dramatic *far way beyond* parts in cycle, can you please spell them out, in a digestible form, like I am trying to do for structuring process with observables/streams.
Do write it out, for me to get it, and for others to be captivated, when they read it on cycle site, or hear on cycle talk.

**Comment by staltz** [comment]
@3n-mb this is a good discussion to have, and it raises some inspiration in me how to write the docs differently. I'm labeling this a discussion because the actionable task here would be to "rewrite the docs", that it's too broad to be considered a normal issue. I intend to give the docs a big renovation.

That said, on one hand I feel like the docs already explain why do we use streams, there is a whole section on that: https://cycle.js.org/streams.html. On the other hand, it's true, the docs page is a bit too technical and assumes a bit too much from the reader. In the future, I think the docs will emphasize more the dataflow/pipeline + I/O (sources/sinks) aspect than the functional and reactive aspect. F&RP is just *how* we accomplish a dataflow paradigm.

**Comment by geovanisouza92** [comment]
@3n-mb 

> If you associate this with cycle only, and project such view onto newbies, then they will not be writing observables-based, or cycle style logic, when they are not using cycle.

Is about that I'm talking about. Cycle.js as an architectural pattern is beyond UI-only processes, and, as you said, newbies could use the pattern outside Cycle.js world. This is one of the biggest possibilities that Cycle.js provides us, unlike many other frameworks.

But, should we assume the responsibility to explain what Streams/Observables are? Is that a main concern for Cycle.js? There is sufficient (and high-quality) material out there, that we could point newcomers to?

We ask too much for newcomers: We expect that (1) they grok what Streams/Obserables are, (2) understand the dialogue between users and computers, (3) all the stream pumbling from one component to another (that is something that I'm trying to think how to turn easier), (4) a big mindset shift, (5) new patterns even to make a HTTP request (there's not action involved, the request and response are described), etc.

This should be the focus: How could we communicate all the "wow" effect that we had when learned all this in an easy and familiar way, getting rid of complicated language, new terms (even when they are the correct ones), rich step-by-step examples and so on.

Just to be clear: I'm not opposing your proposal, actually I liked the idea. :smiley:

**Comment by 3n-mb** [comment]
> We ask too much for newcomers: We expect that (1) they grok what Streams/Obserables are, (2) understand the dialogue between users and computers, (3) all the stream pumbling from one component to another (that is something that I'm trying to think how to turn easier), (4) a big mindset shift, (5) new patterns even to make a HTTP request (there's not action involved, the request and response are described), etc.

Let's stop asking of them too much. Let's move them step-by-step.
Following quoted points:
 1) *Groking what Streams/Obserables are.* This is what we want them to absorb. This is the gold they need to see.
 2) *Dialogue between users and computers.* This is an analogy for trying to describe (1). This is not an additional point for user to learn.
 3) *All the stream pumbling from one component to another.* This is how you shape processes, cast in observables. This is where developer has fun. This is what allows her to make today something, that seemed impossibly complicated yesterday. But it is again part of (1). It is a part of package that answers the question: "Why should I use observables/streams?"
 4) *A big mindset shift.* We should get off the high horse here, alright? No mindset shift needed. These are just loud words. Describe what streams/observables are in terms that developers already use, not in terms collections/enumerations, which they'll naturally get themselves.
 I think of cycle components as objects. Yes. For me cycle is an Object Oriented Design. Main creates a circuitry that is object. This object has a clear boundary (sources, sinks). All of code, process state and data state, logically related to this object, sits in it, cleanly compartmentalized. And remember, original OOP should've been called Message Oriented Programing. Well, events are those messages!
See, no grand *mindset shift* needed here.
 5) *New patterns even to make a HTTP request*. This you should stop. As it has been said elsewhere, when my code makes a request, it logically waits for a reply. Hence, request is a pull action. There are two types of actions, push and pull. And this is an irreducible set.
 As long as cycle sticks with let's turn pull, like http request, into push pattern, devs will be (a) confused, (b) turned off.

Let's recoup. Out of 5 points, we should only do number 1, which naturally includes 3.

**Comment by 3n-mb** [comment]
@staltz in the light of seeing cycle as an object oriented design, expressed [above](https://github.com/cyclejs/cyclejs/issues/693#issuecomment-333667507), and may be elsewhere, it would be cool if you do a talk on it. Yes.

Cycle components are clearly separated. Message passing as events. Your (developer) code is functional, while required state for process' is handled by operators. And in a non-functional world one may write an almost pure functional code.

In one your talks you dropped condescending comments like "no object", "no this", implying a belittling attitude towards OO. That wasn't winning people hearts. It fed narrative that functional programming is somehow opposite of OOP, when OOP done right is what cycle does.
Except for missing pull object, that should exist like second order stream object exists for push. But it is a fixable and non-major caveat.

I'd love to see you amend the divisive incorrect view. As a bonus, it will bring more people to cycle.
Heck. Just like saying cycle is FRP, you could be saying cycle is OOD. And site/docs may have this.
Sometimes you sell with FRP, sometimes you sell with OOD.

**Comment by 3n-mb** [comment]
@staltz @geovanisouza92 
As a reference, related OOD talk is [this](https://www.youtube.com/watch?v=RdE-d_EhzmA)

**Comment by 3n-mb** [comment]
@geovanisouza92 
> But, should we assume the responsibility to explain what Streams/Observables are?

I think, cycle has no choice here. But there is a shortcut.

> Is that a main concern for Cycle.js?

Is adoption a main concern for cycle?

> There is sufficient (and high-quality) material out there, that we could point newcomers to?

Yes and no. Yes, there are docs (English) and nice rxmarbles. But these say *what* observables are.
None of those docs say why should I use them. And cycle should promote the tiny *why* part, while using/directing for *what* parts to existing docs (this is a shortcut).

To add insult to the injury, all tutorials start with "this is difficult ... at first", or "its a totally new way to see the world".
The docs may start from, enumerables and comparison to them. But this is talking past developer.

If observable's docs and tutorials were great, we would not need to onboard people to streams. But we have to do it. Hence, those other docs are not doing their job.

**Comment by geovanisouza92** [comment]
@3n-mb Now I see your points, ad your bring some good ones.

So, you are basically saying that we should invest time on teaching Observables / Streams and also lower the barrier removing complicated terms and this dichotomy that drives users to think that they should throw everything that they learned away and start from scratch?

In that case, I can only agree with you.

---

I think that a glossary could be a nice starting point.

**Comment by 3n-mb** [comment]
@staltz let's go over [https://cycle.js.org/streams.html](https://cycle.js.org/streams.html).

[First part](https://cycle.js.org/streams.html#streams-reactive-programming) describes in great length how to do "Separation of Concerns".
Call this section "Separation of Concerns", and somewhere at the end say that putting arrows in this way, clean way, is what is called reactive programming. At the end, when you have already unpacked the most important *why should I do this?* part.
A dev is told everywhere to do "Separation of Concerns", hence it will be his language. And she will be pleasantly surprised that such cleanliness is *Reactive Programming* (applause, ladies and gentlemen, :clap: ).

First paragraph should be gone. Or change it to something like "we want cycle to help you write a clean code, etc.", as an intro to "Separation of Concerns".

The "*There is a lot of confusion surrounding what Reactive means*" in the first paragraph already introduces an aura of difficulty. Remove it.

Current last paragraph is a no go.
 - "*The challenge with Reactive programming is this paradigm shift where we attempt ...*" this is scaring pants out of reader.
 - "*After rewiring your brain to think ...*" this begs a reply: get off your high horse. This sentence pre-supposses that my (reader's) brain is mis-wired. Mis-wired brain in folk culture is a polite for "dumb".
I know, it was written with good intentions. Just deleted it.

May be say that events in reactive programing can be carried in streams. This will be a take away into next part (which 1st paragraph should be gone), and it seeds explicitly understanding that streams carry events. Events are first order objects, while, one abstraction layer up, streams are second order.
This people will get. And it will be a nice take away into next section and allow to talk about operators, and how these do ton of work for you.

[In second part](https://cycle.js.org/streams.html#streams-what-is-a-stream), instead of "*Instead of teaching RxJS or xstream on this site, we recommend a few great learning resources*" just say that we'd love to copy the following resources here, cause they are so good, but due to format and respect of copyright, we links 'em here.

And sell in "Streams" part the power of operators.

Overall point:
May be this whole session should be called *Why Cycle*, not Streams.

General section of a site can be:
 - Getting started (very short, even shorter)
 - Why Cycle
 - Installation ... advanced getting started.

Put Dialog thingy into MVI. Say philosophy for MVI. Then MVI, Components, drivers.

**Comment by 3n-mb** [comment]
Looking at [https://cycle.js.org/basic-examples.html](https://cycle.js.org/basic-examples.html)

> This is the same “conversation initiative” problem we saw in the previous “checkbox” example.

No. Don't say that anything is a problem, cause cycle should solve my problems, not introduce new ones.
Remove the negative complexity tone.

**Comment by 3n-mb** [comment]
Looking at [https://cycle.js.org/basic-examples.html#basic-examples-increment-a-counter-what-is-the-convention](https://cycle.js.org/basic-examples.html#basic-examples-increment-a-counter-what-is-the-convention)
> But how do we create a count$? Clearly it must depend on increment clicks and decrement clicks.

Add an asterisk with "*sorry kids, there are no templates. You have to do `.select()` like you did in jQuery days, remember? Wasn't it fun?*"

Here we lose 90% of audience.
Just make a trivial library to accommodate me here. It must be trivial. And it doesn't have to be in core. Just a lib. :pray:

**Comment by 3n-mb** [comment]
Without templates lib, you expose me to thinking about raw html. :sos:

**Comment by 3n-mb** [comment]
Templates must be an option. Then you allow for both raw html access and white-gloves template abstraction.

Templates is trivial, cause only sources and sinks cross border. Its OODD.

**Comment by 3n-mb** [comment]
@geovanisouza92 
> Cycle.js (at least @cycle/run*) could be used everywhere to structure asynchronous streams graphs

I pose a challenge. I have an app in MVI architecture. *My code*, consumes sources of events from *intent* section, and it sinks view manipulating events into *view* section. VI uses templates.
Can you make a template library in cycle, so that I can directly move to cycle. Note that you don't need to see code of my app, cause all that is needed, provide sources and sinks to it, based on a template.

This is a challenge. Otherwise, how can you go around saying that cycle is the best, if this simple scenario cannot be implemented?

**Comment by wyqydsyq** [comment]
I'm not sure where the demand for templates comes from @3n-mb, what makes regular hyperscript / JSX insufficient? I really think templates are just something that's "nice to have" and not an essential part of a UI framework. React is the most popular UI framework and even they don't have any concept of templates, just regular components (albeit often OOP class-based instead of pure functions like in Cycle).

If you really need support for a static template language like Handlebars etc, you could create a Driver wrapping `@cycle/DOM`  that compiles your templates into a registry of template functions (from what I've seen most template libraries compile the template string into a function that accepts inputs and returns a HTML string in a similar manner to [EJS](http://ejs.co/)), when a component emits on your driver's sink, the driver passes that payload to the respective template function, giving us HTML which can then convert that HTML string to a VTree using [snabbdom-virtualize](https://github.com/appcues/snabbdom-virtualize), which can be passed off to `snabbdom` for DOM mounting, diffing, etc.

**Comment by geovanisouza92** [comment]
@wyqydsyq said exactly what I was thinking. I'm not saying that template language is not important, but it's nice to have, as said.

And if you think that's an important missing feature, **I challenge you** to write an experiment and propose to the community. Cycle.js is full of small tools that help here and there, and eventually some go to the 

---

When I say that Cycle.js is an architectural pattern, I'm talking about this idea of connecting streams of events and separate logic from effects, plugging them using "run". It's not tied for UI only, but any source of events. This is a broader understanding of Cycle.js and this doesn't need to be said to new users, because eventually they will get it.

**Comment by whitecolor** [comment]
> Add an asterisk with "sorry kids, there are no templates. You have to do .select() like you did in jQuery days, remember? Wasn't it fun?"

> Templates must be an option. Then you allow for both raw html access and white-gloves template abstraction.

Don't get what you mean by "templates". One should use `.select()` is used because of completely unidirectional nature of cycle, don't get why it is should be compared to jquery, it is just externally alike API, the essence is different.

**Comment by staltz** [comment]
> First paragraph should be gone. Or change it to something like "we want cycle to help you write a clean code, etc.", as an intro to "Separation of Concerns".

@3n-mb You have quite a lot of good insights about the docs, I think it would be worth submitting a few PRs. Make each PR small and it will maximize its chance of being merged quickly and easily. This is how we go forwards. :)

**Comment by 3n-mb** [comment]
@whitecolor 

> why it should be compared to jquery

Cause I haven't touched raw dom since jQuery times. Cause I haven't put ids on elements to find them later since jQuery times.

---
@wyqydsyq 
> what makes regular hyperscript / JSX insufficient?

Short answer: `.select()`.
My understanding is that it will be required together with putting `id`'s on tags.

This is raw dom for my senses. This is jQuery past.

---
@geovanisouza92 
> I challenge you to write an experiment and propose to the community.

I was already on **this challenge**. #631 Unfortunately, @staltz didn't look deeper, and incorrectly said that it is #501 . #631 is a liberating, smoothest path way.

So, I rephrased it for myself into doing my code in observables, while taking VI sources/sink from other UI library that has templates (and doesn't require ids, selects). And I succeeded, and I want to share that structuring process with observables/streams is the most important thing here, while a form of main, selects, are mere options, of which there may be a variety on the menu.

---
@whitecolor @geovanisouza92 
> unidirectional nature of cycle

It is a flow of events, **first order** objects, that is unidirectional.
It shouldn't restrict shape of function that builds streams **second order** objects.

I guess cycle needs ids to maintain particular shape of main. I am saying that by providing an alternative option for main doesn't change unidirectional nature of **first order** objects' flow.

Note: I mentioned this in #631 , albeit in different words.

---
@staltz 

> Make each PR small and it will maximize its chance of being merged quickly and easily.

Maximum efficiency can be attained, when I speak to you cause, quote "*After rewiring your brain to think*" in a new way all problems will fix themselves.

...
Just in case, it has not been felt properly, I'll say it explicitly. I expected a different reaction, when I pointed that you put an insult in docs.

**Comment by whitecolor** [comment]
> My understanding is that it will be required together with putting id's on tags.

> This is raw dom for my senses. This is jQuery past.

You proposed something like this:

```js
function main(sources) {
  const decrementUserAction = xs.subject();
  const decrement$ = decrementUserAction.mapTo(-1);

  const incrementUserAction = xs.subject();
  const increment$ = incrementUserAction.mapTo(+1);

  const action$ = xs.merge(decrement$, increment$);
  const count$ = action$.fold((x, y) => x + y, 0);

  const vdom$ = count$.map(count =>
    div([
      button('Decrement', { click: decrementUserAction }),
      button('Increment', { click: incrementUserAction }),
      p('Counter: ' + count)
    ])
  );
  return { DOM: vdom$ };
}
```

Not much simplier than:

```js
function main(sources) {
  const decrementUserAction = cuid(); // random Id, can be used constant whatever
  const decrement$ = DOM.select(`.{decrementUserAction}`).events('click')mapTo(-1);

  const incrementUserAction = cuid();
  const increment$ = DOM.select(`.{incrementUserAction}`).events('click').mapTo(+1);
  ...
  const vdom$ = count$.map(count =>
    div([
      button(`.{decrementUserAction}`, 'Decrement'),
      button(`.{incrementUserAction}`, 'Increment'),
      p('Counter: ' + count)
    ])
  );
  return { DOM: vdom$ };
}
```

Or if you like, you may even implement your variant with streams passed to sink using snabbdom events module.

```js
function main(sources) {
  const decrementUserAction = xs.create();
  const decrement$ = decrementUserAction.mapTo(-1);

  const incrementUserAction = xs.create();
  const increment$ = incrementUserAction.mapTo(+1);

  const action$ = xs.merge(decrement$, increment$);
  const count$ = action$.fold((x, y) => x + y, 0);

  const vdom$ = count$.map(count =>
    div([
      button('Decrement', { on: {
        click: (ev) = decrementUserAction.shamelesslySendNext(ev)
      } }),
      ....
      p('Counter: ' + count)
    ])
  );
  return { DOM: vdom$ };
}
```

**Comment by 3n-mb** [comment]
@whitecolor exactly! It is a trivial issue.
It can be used with drivers, or even allow for setting without drivers: driver thingy will spin only in template engine, developer won't have to know about drivers at all (better onboarding :1st_place_medal: ).
> Or if you like, you may even implement your variant with streams passed to sink using snabbdom events module.

I don't believe my PR will be accepted. :smiling_imp: Even issue #631 was killed fast.

All I can hope for is that someone else notices that there are user like me. Ya, its that overwhelming majority that does angular, vue, aurelia, cause they provide templates.

**Comment by 3n-mb** [comment]
I think I question the wisdom of making `main` unidirectional, while

> >   unidirectional nature of cycle

> It is a flow of events, first order objects, that is unidirectional.
It shouldn't restrict shape of function that builds streams second order objects.

Clean description of what events and streams are only surfaces a confusion among you, folks.

You have to decide for yourself, if a particular form of `main` (now it restricts templates, and needs semi globals, opposite to functional spirit, ref: #631 ) is the golden nugget.

When you do such decision, put it up. So that when #631 like things are short down, you clearly say that it is done for this obscure constraint.

**Comment by whitecolor** [comment]
> I don't believe my PR will be accepted. 😈 Even issue #631 was killed fast.

I mean you can implement as custom snabbdom module, if you like this way so much.

**Comment by geovanisouza92** [comment]
@3n-mb Just to start arguing your points: You could have exactly that "template-ish" kind of DOM by creating a module for snabbdom and [providing it](https://github.com/cyclejs/cyclejs/blob/112d1b8be6c8f79a0c467a5eb0f80192185157b6/dom/src/makeDOMDriver.ts#L55) for the current DOM driver

**Comment by 3n-mb** [comment]
@whitecolor 
> I mean you can implement as custom snabbdom module, if you like this way so much.

Exactly, I already did it for my project, except that instead of snabbdom I chose AngularJS with Material Design.
... What drivers? It all is abstracted away, thanks to structuring process with observables/streams.

---

I also don't feel like using shameful functions in API. Insult in docs goes hand-in-hand with adding emotionally charged naming in API. Won't you agree?
